### PR TITLE
feat: Raster App

### PR DESCRIPTION
**Pull request by emwp**
## Purpose
Raster enables users to leverage their existing media assets and easily use them in their Contentful projects.

## Approach
Field View:
image

Dialog View:
image

## Testing steps
1. Install the app.
2. Include JSON Object field on content model.
3. Change the above field appearance to use Raster.
4. Create an entry and test the Raster field – You can select a single image or multiple images from different libraries.

## Breaking Changes
No breaking changes.

### CODE REVIEW

**Comment by lambardi** [comment]
Thanks! We'll get a CR started ASAP.

**Comment by emwp** [comment]
> Thanks! We'll get a CR started ASAP.

Hey @lambardi, thanks!
One thing I wanted to make, our app is a Next.js app, I believe apps need to be hosted by Contentful, right?
Can Contentful host the Next.js app? I ask because I noticed all other apps are using CRA/Vite.

**Review by whitelisab — CHANGES_REQUESTED** [review]
We added some comments for you to review and address. In addition to those comments, here are few other things to consider:

- Please add a license file for your app. You can review other apps in this repo as a reference (e.g.: https://github.com/contentful/marketplace-partner-apps/blob/main/apps/wix/LICENSE). 
- Confirm that your app adheres to OWASP standards: https://owasp.org/www-project-top-ten/
- Confirm that your app meets basic accessibility standards: https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML

File: apps/raster/package.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
+{
+  "name": "raster",
+  "version": "1.0.0",
```
[the remark this answers is not in the document]
Comment (emwp) [reply apps/raster/package.json]: Done!

File: apps/raster/package.json [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  "private": true,
+  "scripts": {
```
[the remark this answers is not in the document]
Comment (emwp) [reply apps/raster/package.json:5]: Sure! I've added some testing for the ConfigScreen, Field and Dialog components.

File: apps/raster/package.json [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    "dev": "next dev",
+    "build": "next build",
+    "start": "next start",
+    "lint": "next lint",
+    "install-ci": "npm ci",
+    "create-app-definition": "contentful-app-scripts create-app-definition",
+    "deploy": "contentful-app-scripts upload --ci --bundle-dir ./dist --organization-id ${DEFINITIONS_ORG_ID} --definition-id 41ISJmeSvLGsYT60RfIqyd --token ${CONTENTFUL_CMA_TOKEN}"
```
[the remark this answers is not in the document]
Comment (emwp) [reply apps/raster/package.json]: Done!

File: apps/raster/src/components/locations/ConfigScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+export interface AppInstallationParameters {
+  orgId: string | undefined;
+  apiKey: string | undefined;
+}
+
+function ConfigScreen() {
+  const [parameters, setParameters] = useState<AppInstallationParameters>({
+    orgId: "",
+    apiKey: "",
```
[the remark this answers is not in the document]
Comment (emwp) [reply apps/raster/src/components/locations/ConfigScreen.tsx]: The `apiKey` from Raster is for read only, not necessarily a secret.

**Comment by emwp** [comment]
@whitelisab Seem to be having some issue with node/npm version.
That's why I had removed the eslint config before haha, but I've tried with many node versions that should be on the accepted range and still getting this issue. Also see that other apps have the same dependency installed with no issues.

https://app.circleci.com/pipelines/github/contentful/marketplace-partner-apps/8553/workflows/13d9b627-d21f-44cf-a3f3-91f62c61c86f/jobs/7053

**Comment by t-col** [comment]
hi @emwp, I'm taking over for @whitelisab here. Regarding the CI issue with node versions: if you rebase you should see that go away. I believe the problem was with the node version that we had pinned in the root `.npmrc`, which I fixed [here](https://github.com/contentful/marketplace-partner-apps/pull/2521).

**Review by t-col — CHANGES_REQUESTED** [review]
@emwp looks like we're just missing a LICENSE file. Once we have that I can get this merged. See [this one](https://github.com/contentful/marketplace-partner-apps/blob/main/apps/wix/LICENSE) as a reference

**Comment by emwp** [comment]
> @emwp looks like we're just missing a LICENSE file. Once we have that I can get this merged. See [this one](https://github.com/contentful/marketplace-partner-apps/blob/main/apps/wix/LICENSE) as a reference

Thanks @t-col, License is added.

**Review by t-col — APPROVED** [review]

**Comment by t-col** [comment]
hi @emwp - can you confirm that you were able to upload your app (Contentful hosted, as opposed to served locally) to a test org and install it to a space/env successfully? I'm unable to do so for the production deployment.

**Comment by emwp** [comment]
> hi @emwp - can you confirm that you were able to upload your app (Contentful hosted, as opposed to served locally) to a test org and install it to a space/env successfully? I'm unable to do so for the production deployment - the app does not render at all, and there appears to be an issue fetching the built assets from the HTML file (maybe a CORS issue). I unsuccessfully attempted to tweak the next.js build config to resolve the issues I'm seeing, but no luck. Please let me know if you experience this issue when you build and upload the app on your end.

Hey @t-col – I can't confirm if I did exactly that, what I did was with the build folder (where it says to drag it to on Contentful).
I believe the best solution is to just migrate this to a CRA/Vite app, we started as a Next.js app just by default without knowing how the deployment process would be for Contentful.

I've already migrated most of it to Vite app, will ping you here when I open a PR for this.

**Comment by t-col** [comment]
hey @emwp we use a utility from `contentful-app-scripts` that looks exactly like this: https://github.com/contentful/create-contentful-app/blob/main/packages/contentful--app-scripts/README.md#non-interactive-mode
Use this as a reference in uploading to your own space.

**Comment by emwp** [comment]
> hi @emwp - can you confirm that you were able to upload your app (Contentful hosted, as opposed to served locally) to a test org and install it to a space/env successfully? I'm unable to do so for the production deployment - the app does not render at all, and there appears to be an issue fetching the built assets from the HTML file (maybe a CORS issue). I unsuccessfully attempted to tweak the next.js build config to resolve the issues I'm seeing, but no luck. Please let me know if you experience this issue when you build and upload the app on your end.

Just added the PR here if you can take a look:
https://github.com/contentful/marketplace-partner-apps/pull/2699

**Pull request merged**